core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -929,32 +938,51 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
       inode = inodePath.getInode();
       auditContext.setSrcInode(inode);
 
-      List<FileInfo> ret = new ArrayList<>();
-      if (inode.isDirectory()) {
-        TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath);
-        try {
-          mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, inodePath);
-        } catch (AccessControlException e) {
-          auditContext.setAllowed(false);
+      List<FileInfo> ret = listStatusInternal(inodePath, auditContext,
+          listStatusOptions.isRecursive(), inodePath);
+
+      auditContext.setSucceeded(true);
+      Metrics.FILE_INFOS_GOT.inc();
+      return ret;
+    }
+  }
+
+  private List<FileInfo> listStatusInternal(LockedInodePath currInodePath,
+                                                  AuditContext auditContext,
+                                                  boolean recursive,
+                                                  LockedInodePath originalInodePath)
+      throws FileDoesNotExistException, UnavailableException, AccessControlException {
+    List<FileInfo> statusList = new ArrayList<>();
```

Comment:
I see. Since now `listStatusInternal` is recursive, maybe we should pass in the mutable list so the recursion can just add to a single list. Constructing many lists sounds bad. What do you think?

Reply:
Yeah, I agree.